mayastor/src/bdev/nexus/nexus_bdev.rs

```diff
@@ -1058,61 +1071,40 @@ pub async fn nexus_create(
     // closing a child assumes that the nexus to which it belongs will appear
     // in the global list of nexus instances. We must also ensure that the
     // nexus instance gets removed from the global list if an error occurs.
```

Comment:
This comment still applies.

We can now potentially call Nexus::close_children() from within Nexus::new().
NexusChild::close() calls NexusChild::destroy() which raises a SPDK_BDEV_EVENT_REMOVE event.
This ultimately results in callback core::bdev::event_cb being called, which then calls lookup_child_from_bdev().
This last function only works correctly if the owning nexus is present in the global list of nexus instances.
However, it is now currently not present (it's only added at the end), and the bdev will not be properly closed/destroyed.

We must either change the way lookup_child_from_bdev() works to avoid this issue, or otherwise ensure that the current nexus is present in the global list by the time Nexus::close_children() is called, otherwise mayastor will hang.

Reply:
Just tested on this branch by running
```
/mayastor-client nexus create $id 4Mib 'malloc:///m0?size_mb=4'
``` 
and the client hangs,

The last thing you see in the log is the following:
```
[2021-03-08T10:26:46.504251499+00:00  WARN mayastor::bdev::nexus::nexus_bdev:nexus_bdev.rs:514] error updating child labels: Device is too small to accomodate Metadata partition: blocks=8192
[2021-03-08T10:26:46.505886123+00:00 ERROR mayastor::bdev::nexus::nexus_bdev:nexus_bdev.rs:1092] failed to open nexus nexus-82520992-f822-47d1-93bc-fd5345334d9d: Failed to read child label of nexus nexus-82520992-f822-47d1-93bc-fd5345334d9d: Label is invalid: Incorrect MBR signature
[2021-03-08T10:26:46.506056576+00:00  INFO mayastor::bdev::nexus::nexus_child:nexus_child.rs:361] Closing child malloc:///m0?size_mb=4
[2021-03-08T10:26:46.506211046+00:00  INFO mayastor::nexus_uri:bdev_destroy{uri="malloc:///m0?size_mb=4"}:nexus_uri.rs:79] new
[2021-03-08T10:26:46.506320282+00:00  INFO mayastor::nexus_uri:bdev_destroy{uri="malloc:///m0?size_mb=4"}:nexus_uri.rs:79] enter
[2021-03-08T10:26:46.506628106+00:00  INFO mayastor::nexus_uri:bdev_destroy{uri="malloc:///m0?size_mb=4"}:nexus_uri.rs:79] exit
[2021-03-08T10:26:46.506755874+00:00  INFO mayastor::core::bdev:bdev.rs:168] Received remove event for bdev m0
```
Basically the remove event is not being handled correctly and spdk is stuck in the destroy code for the bdev.

Reply:
Hmm so it seems `Nexus::create` could return an `Rc<Nexus>` instead of a `Box<Nexus>` since it needs to be present in the global state to complete child registration? Otherwise I guess it would have to add then remove itself from the list? :thinking: 

Should `Nexus::new()` populate the global list? If not, then this PR should be closed I guess.

Reply:
Oh, I noticed this problem even occurs on `develop`... Hm...

Reply:
Actually the more I think about this, the more I'm inclined to take the view that the best solution at present is to remove the "child_bdevs" argument from Nexus::new() and dis-allow adding children like this altogether. I would also go as far as saying that Nexus::new() probably should not be public either. What do others think?

Reply:
So how is it declarative if you have to build a nexus by iterating over the children?

How would `Nexus::new` be called then? It seems like this is intended to be public API? It's certainly more idiomatic than `nexus_create`, which has no business being in public API.

Reply:
As far as I can tell, a proper and correct implementation of a public new() would need to do more or less what nexus_create() does. So perhaps we should retain the existing new() as a (private) utility function appropriately renamed, make the existing nexus_create() our new(),  and nexus_create() becomes a simple wrapper around new(). Also, this would make nexus_instance_new() superflous as the nexus needs to be added to the list of global instances as part of new()